src/driver-gskill.c

```diff
@@ -0,0 +1,1365 @@
+{
+	const struct gskill_button_function_mapping *mapping;
+
+	ARRAY_FOR_EACH(gskill_button_function_mapping, mapping) {
+		if (mapping->type == type)
+			return &mapping->action;
+	}
+
+	return NULL;
+}
+
+static enum gskill_button_function_type
+gskill_button_function_from_action(const struct ratbag_button_action *action)
+{
+	const struct gskill_button_function_mapping *mapping;
+
+	ARRAY_FOR_EACH(gskill_button_function_mapping, mapping) {
+		if (ratbag_button_action_match(&mapping->action, action))
+			return mapping->type;
+	}
+
+	return GSKILL_BUTTON_FUNCTION_DISABLE;
+}
+
+static uint8_t
+gskill_calculate_checksum(const uint8_t *buf, size_t len)
+{
+	uint8_t checksum = 0;
+	unsigned i;
+
+	for (i = GSKILL_CHECKSUM_OFFSET + 1; i < len; i++)
+		checksum += buf[i];
+
+	checksum = ~checksum + 1;
+
+	return checksum;
+}
+
+static int
+gskill_general_cmd(struct ratbag_device *device,
+		   uint8_t buf[GSKILL_REPORT_SIZE_CMD]) {
+	int rc;
+	int retries;
+
+	rc = ratbag_hidraw_raw_request(device, GSKILL_GENERAL_CMD, buf,
+				       GSKILL_REPORT_SIZE_CMD,
+				       HID_FEATURE_REPORT, HID_REQ_SET_REPORT);
+	if (rc < 0) {
+		log_error(device->ratbag,
+			  "Error while sending command to mouse: %d\n", rc);
+		return rc;
+	}
+
+	for (retries = 0; retries < 10; retries++) {
+		/* Wait for the device to be ready */
+		msleep(20);
+
+		rc = ratbag_hidraw_raw_request(device, 0, buf,
+					       GSKILL_REPORT_SIZE_CMD,
+					       HID_FEATURE_REPORT,
+					       HID_REQ_GET_REPORT);
+		/*
+		 * Sometimes the mouse just doesn't send anything when it wants
+		 * to tell us it's ready
+		 */
+		if (rc == 0) {
+			continue;
+		} else if (rc < GSKILL_REPORT_SIZE_CMD) {
+			log_error(device->ratbag,
+				  "Error while getting command response from mouse: %d\n",
+				  rc);
+
+			return rc;
+		}
+
+		/* Check the command status bit */
+		switch (buf[1]) {
+		case 0: /* sometimes the mouse gets lazy and just returns a
+			   blank buffer on success */
+		case GSKILL_CMD_SUCCESS:
+			return 0;
+
+		case GSKILL_CMD_IN_PROGRESS:
+			break;
+
+		case GSKILL_CMD_IDLE:
+			log_error(device->ratbag,
+				  "Command response indicates idle status? Uh huh.\n");
+			return GSKILL_CMD_IDLE;
+
+		case GSKILL_CMD_FAILURE:
+			log_error(device->ratbag, "Command failed\n");
+			return GSKILL_CMD_FAILURE;
+
+		default:
+			log_error(device->ratbag,
+				  "Received unknown command status from mouse: 0x%x\n",
+				  buf[1]);
+			return -1;
+		}
+	}
+
+	log_error(device->ratbag,
+		  "Failed to get command response from mouse after %d tries, giving up\n",
+		  retries);
+	return -1;
+}
+
+static int
+gskill_get_active_profile_idx(struct ratbag_device *device)
+{
+	uint8_t buf[GSKILL_REPORT_SIZE_CMD] = { 0x0c, 0xc4, 0x7, 0x0, 0x1 };
+	int rc;
+
+	rc = gskill_general_cmd(device, buf);
+	if (rc) {
+		log_error(device->ratbag,
+			  "Error while getting active profile number from mouse: %d\n",
+			  rc);
+		return -1;
```

Comment:
fwiw, I'm just going to add a
```c
if (rc > 0)
    rc = -EPROTO;
```

check in `gskill_general_cmd()` so we can just safely return `rc` whenever `gskill_general_cmd()` fails